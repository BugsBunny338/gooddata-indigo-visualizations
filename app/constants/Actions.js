export const BOOTSTRAP = 'BOOTSTRAP';
export const BOOTSTRAP_DATA = 'BOOTSTRAP_DATA';
export const BOOTSTRAP_ERROR = 'BOOTSTRAP_ERROR';

export const LOGOUT = 'LOGOUT';
export const LOGOUT_DATA = 'LOGOUT_DATA';
export const LOGOUT_ERROR = 'LOGOUT_ERROR';

export const CATALOGUE_UPDATE = 'CATALOGUE_UPDATE';
export const CATALOGUE_UPDATED = 'CATALOGUE_UPDATED';
export const CATALOGUE_SET_ACTIVE_FILTER_INDEX = 'CATALOGUE_SET_ACTIVE_FILTER_INDEX';
export const CATALOGUE_SET_QUERY = 'CATALOGUE_SET_QUERY';
export const CATALOGUE_SET_ACTIVE_DATASET_ID = 'CATALOGUE_SET_ACTIVE_DATASET_ID';

export const DATE_DIMENSIONS_DATA = 'DATE_DIMENSIONS_DATA';

export const DATASETS_REQUEST = 'DATASETS_REQUEST';
export const DATASETS_DATA = 'DATASETS_DATA';

export const ITEM_DETAIL_REQUEST = 'ITEM_DETAIL_REQUEST';
export const ITEM_DETAIL_DATA = 'ITEM_DETAIL_DATA';

export const BUCKETS_SELECT_VISUALIZATION_TYPE = 'BUCKETS_SELECT_VISUALIZATION_TYPE';
export const BUCKETS_SET_BUCKET_ITEM_COLLAPSED = 'BUCKETS_SET_BUCKET_ITEM_COLLAPSED';
export const BUCKETS_SET_BUCKET_ITEM_SHOW_IN_PERCENT = 'BUCKETS_SET_BUCKET_ITEM_SHOW_IN_PERCENT';
export const BUCKETS_SET_BUCKET_ITEM_SHOW_POP = 'BUCKETS_SET_BUCKET_ITEM_SHOW_POP';
export const BUCKETS_SET_BUCKET_ITEM_ADD_FILTER = 'BUCKETS_SET_BUCKET_ITEM_ADD_FILTER';
export const BUCKETS_SET_BUCKET_ITEM_REMOVE_FILTER = 'BUCKETS_SET_BUCKET_ITEM_REMOVE_FILTER';
export const BUCKETS_SET_BUCKET_ITEM_AGGREGATION = 'BUCKETS_SET_BUCKET_ITEM_AGGREGATION';
export const BUCKETS_SET_BUCKET_ITEM_DIMENSION = 'BUCKETS_SET_BUCKET_ITEM_DIMENSION';
export const BUCKETS_SET_BUCKET_ITEM_GRANULARITY = 'BUCKETS_SET_BUCKET_ITEM_GRANULARITY';
