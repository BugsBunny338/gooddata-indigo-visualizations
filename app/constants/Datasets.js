export const ALL_DATA_TRANSLATIONS_KEY = 'datasets.production_data';
export const ALL_DATA_ID = null;
