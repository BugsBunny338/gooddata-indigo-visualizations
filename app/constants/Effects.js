export const REDIRECTION = 'REDIRECTION';
export const SET_HASH = 'SET_HASH';
