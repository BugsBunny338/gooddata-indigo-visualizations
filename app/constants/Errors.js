export const LOAD_ERROR = 'LOAD_ERROR';
export const NOT_AUTHORIZED_ERROR = 'NOT_AUTHORIZED_ERROR';
