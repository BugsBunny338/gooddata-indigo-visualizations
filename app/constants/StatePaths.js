export const BOOTSTRAP = ['bootstrapData'];

export const USER_FIRST_NAME = ['bootstrapData', 'accountSetting', 'firstName'];
export const USER_LAST_NAME = ['bootstrapData', 'accountSetting', 'lastName'];
export const PROFILE_SETTING_SEPARATORS = ['bootstrapData', 'profileSetting', 'separators'];
export const USER_EMAIL = ['bootstrapData', 'accountSetting', 'email'];
export const USER_URI = ['bootstrapData', 'accountSetting', 'links', 'self'];
export const USER_LOGIN_MD5 = ['bootstrapData', 'accountSetting', 'loginMD5'];
export const BOOTSTRAP_DATA_PERMISSIONS = ['bootstrapData', 'permissions'];

export const Permissions = {
    CAN_ACCESS_WORKBENCH: 'canAccessWorkbench',
    CAN_MANAGE_PROJECT_DASHBOARD: 'canManageProjectDashboard'
};

export const PROJECT_TEMPLATE = ['bootstrapData', 'project', 'template'];
export const PROJECT_TITLE = ['bootstrapData', 'project', 'title'];
export const PROJECT_URI = ['bootstrapData', 'project', 'uri'];

export const BRANDING = ['bootstrapData', 'branding'];
export const APPLE_TOUCH_ICON_URL = [...BRANDING, 'appleTouchIconUrl'];
export const IS_BRANDED = [...BRANDING, 'isBranded'];
export const ORGANIZATION_NAME = [...BRANDING, 'organizationName'];
export const FAVICON_URL = [...BRANDING, 'faviconUrl'];
export const APPLICATION_TITLE = [...BRANDING, 'applicationTitle'];

export const MENU_ITEMS = ['header', 'menuItems'];
export const ACCOUNT_MENU_ITEMS = ['header', 'accountMenuItems'];

export const PAGE_TITLE = ['pageTitle'];

export const DEVICE_VIEWPORT = ['meta', 'device', 'viewport'];
export const DEVICE_PIXEL_RATIO = ['meta', 'device', 'pixelRatio'];
export const IS_MOBILE_DEVICE = ['meta', 'device', 'isMobile'];
