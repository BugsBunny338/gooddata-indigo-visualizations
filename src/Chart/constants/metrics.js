export const DEFAULT_FORMAT = '#,##0.00';
