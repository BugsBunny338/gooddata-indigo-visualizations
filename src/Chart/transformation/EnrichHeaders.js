import { get } from 'lodash';
import { DEFAULT_FORMAT } from '../constants/metrics';

export function getMetricNamesValuesHeaderItems(headers, metrics) {
    // Need to transform the headers to match the rawData transformation
    return [{
        id: 'metricNames',
        title: 'Metric',
        type: 'attrLabel',
        uri: '/metricGroup',
        metrics
    }, {
        id: 'metricValues',
        type: 'metric',
        uri: '/metricValues',

        // For axis formatting, use first metric's formatting.
        // For individual series tooltips and labels, formats are
        // handled separately in metricNames
        format: get(metrics, '0.header.format', DEFAULT_FORMAT),

        // use first metric title as default
        title: get(metrics, '0.header.title', '')
    }];
}

function addMetricNamesValues(headers, metrics) {
    return {
        headers: [...headers, ...getMetricNamesValuesHeaderItems(headers, metrics)],
        metrics
    };
}

function splitHeaders(headers) {
    const metrics = [];
    const otherHeaders = [];

    headers.forEach((header, i) => {
        if (header.type === 'metric') {
            metrics.push({
                index: i,
                header
            });
        } else {
            otherHeaders.push(header);
        }
    });

    return {
        headers: otherHeaders,
        metrics
    };
}

// TODO: Cover all the module with tests
export function enrichHeaders(_headers) {
    const { headers, metrics } = splitHeaders(_headers);

    return addMetricNamesValues(headers, metrics);
}
