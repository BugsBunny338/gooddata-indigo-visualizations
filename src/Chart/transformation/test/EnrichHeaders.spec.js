import { enrichHeaders, getMetricNamesValuesHeaderItems } from '../EnrichHeaders';
import { DEFAULT_FORMAT } from '../../constants/metrics';

describe('splitting headers', () => {
    const headers = [{
        id: 'id1',
        title: 'Probability',
        type: 'metric'
    }, {
        id: 'id2',
        title: 'Revenue',
        type: 'metric'
    }, {
        id: 'id2',
        title: 'Opp Name',
        type: 'attrLabel'
    }, {
        id: 'id3',
        title: 'Stage',
        type: 'attrLabel'
    }];

    it('should split headers to metrics', () => {
        const metrics = enrichHeaders(headers).metrics;
        expect(metrics).toHaveLength(2);
        expect(metrics[0]).toEqual({ index: 0, header: headers[0] });
        expect(metrics[1]).toEqual({ index: 1, header: headers[1] });
    });

    it('should split headers to the rest', () => {
        const newHeaders = enrichHeaders(headers).headers;

        expect(newHeaders).toMatchSnapshot();
    });

    describe('getMetricNamesValuesHeaderItems', () => {
        it('adds metricNames and metricValues headers', () => {
            const headersWithoutMetrics = [{
                id: 'metricNames',
                title: 'Metric',
                type: 'attrLabel',
                uri: '/metricGroup',
                metrics: []
            }, {
                id: 'metricValues',
                type: 'metric',
                uri: '/metricValues',
                format: DEFAULT_FORMAT,
                title: ''
            }];
            const actual = getMetricNamesValuesHeaderItems([], []);
            expect(actual).toEqual(headersWithoutMetrics);
        });

        it('uses first metric format when metric present', () => {
            const metrics = [{
                header: {
                    title: 'Metric 1',
                    format: '00##00'
                }
            }];
            const mockHeaders = [{
                id: 'metricNames',
                title: 'Metric',
                type: 'attrLabel',
                uri: '/metricGroup',
                metrics
            }, {
                id: 'metricValues',
                type: 'metric',
                uri: '/metricValues',
                format: '00##00',
                title: 'Metric 1'
            }];
            const actual = getMetricNamesValuesHeaderItems([], metrics);
            expect(actual).toEqual(mockHeaders);
        });
    });
});
