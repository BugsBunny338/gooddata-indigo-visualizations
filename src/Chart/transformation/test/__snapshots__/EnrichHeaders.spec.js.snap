// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`splitting headers should split headers to the rest 1`] = `
Array [
  Object {
    "id": "id2",
    "title": "Opp Name",
    "type": "attrLabel",
  },
  Object {
    "id": "id3",
    "title": "Stage",
    "type": "attrLabel",
  },
  Object {
    "id": "metricNames",
    "metrics": Array [
      Object {
        "header": Object {
          "id": "id1",
          "title": "Probability",
          "type": "metric",
        },
        "index": 0,
      },
      Object {
        "header": Object {
          "id": "id2",
          "title": "Revenue",
          "type": "metric",
        },
        "index": 1,
      },
    ],
    "title": "Metric",
    "type": "attrLabel",
    "uri": "/metricGroup",
  },
  Object {
    "format": "#,##0.00",
    "id": "metricValues",
    "title": "Probability",
    "type": "metric",
    "uri": "/metricValues",
  },
]
`;
